frontend/src/concepts/notificationPoller/NotificationPollerContext.tsx

```diff
@@ -0,0 +1,98 @@
+import * as React from 'react';
+import { AppNotification } from '~/redux/types';
+import useNotification from '~/utilities/useNotification';
+
+export type NotificationPollerContextType = {
+  watchForNotification: (item: NotificationPollerItem) => Promise<void>;
+};
+
+type NotificationPollerContextProviderProps = {
+  children: React.ReactNode;
+};
+
+export type FinalNotificationPollerResponse =
+  | ({ status: 'success' | 'error' } & Pick<AppNotification, 'title' | 'message' | 'actions'>)
+  | { status: 'stop' };
+
+export type NotificationPollerResponse = FinalNotificationPollerResponse | { status: 'repoll' };
+
+export type NotificationCallback = (signal: AbortSignal) => Promise<NotificationPollerResponse>;
+
+export type NotificationPollerItem = {
+  delayRepollMs?: number;
+  callback: NotificationCallback;
+};
+
+export const NotificationPollerContext = React.createContext<NotificationPollerContextType>({
+  watchForNotification: () => Promise.resolve(),
+});
+
+const DEFAULT_REPOLL_INTERVAL_MS = 0;
+
+export const NotificationPollerContextProvider: React.FC<
+  NotificationPollerContextProviderProps
+> = ({ children }) => {
+  const notification = useNotification();
+  const abortControllersMapRef = React.useRef(new Map<NotificationPollerItem, AbortController>());
+  const timeoutIdsMapRef = React.useRef(
+    new Map<NotificationPollerItem, ReturnType<typeof setTimeout>>(),
+  );
+
+  const invoke = React.useCallback(
+    async (itemToInvoke: NotificationPollerItem, signal: AbortSignal) => {
+      const delayRepollMs = itemToInvoke.delayRepollMs ?? DEFAULT_REPOLL_INTERVAL_MS;
+      const response = await itemToInvoke.callback(signal);
+
+      if (signal.aborted) {
+        return;
+      }
+
+      if (response.status === 'repoll') {
+        const timeoutId = setTimeout(async () => {
+          timeoutIdsMapRef.current.delete(itemToInvoke);
+          await invoke(itemToInvoke, signal);
+        }, delayRepollMs);
+
+        timeoutIdsMapRef.current.set(itemToInvoke, timeoutId);
+        return;
+      }
+
+      abortControllersMapRef.current.delete(itemToInvoke);
+
+      if (response.status === 'success') {
+        notification.success(response.title, response.message, response.actions);
+      } else if (response.status === 'error') {
+        notification.error(response.title, response.message, response.actions);
+      }
+    },
```

Comment:
I think you're opening yourself up to if we add another status it falls into the void... which you're currently using to describe `stop`.

We don't do `exhaustiveChecks` yet -- but I feel maybe we should in this case. It seems it's not a first class of TS, but something we might need to code it up. I found several blogs about it... but I can't find anything in TS 😞

Reply:
Oh nope... I'm just bad at reading Google after they changed their website -- TypeScript Book describes this: https://gibbok.github.io/typescript-book/book/exhaustiveness-checking/

Reply:
TBH I wasn't sure if we should do something on the `stop` state other than nothing. I guess I'd let the code evolve naturally if we ever needed to add or handle other states. Anyway, let me add the structure you've mentioned.

Reply:
Yeah -- the way you have coded it feels like stop would be handled by doing nothing lol... but lets make sure we get TS in a place to help us out so accidents don't happen later 🙂 We have it, might as well use it for all its worth.

Reply:
Interestingly, TS complains about this type setup (as opposed to the example in the book) on the default block because it is sure that code will never reach there. Either way, a switch block looked prettier.

Reply:
Oops, missed seeing this... I added another comment to help with how this can help: https://github.com/opendatahub-io/odh-dashboard/pull/3769#discussion_r1962137347